### ISSUE TITLE
The method 'detach' isn't defined for the class 'FocusScopeNode'

### ISSUE THREAD

**Issue by AndreaMiotto** [flutter/flutter#31784]
After a flutter upgrade and a packages upgrade I'm having this error:

**The method 'detach' isn't defined for the class 'FocusScopeNode'** on:

`FocusScope.of(context).detach();
`

> [✓] Flutter (Channel master, v1.5.9-pre.46, on Mac OS X 10.14.4 18E226, locale it-IT)
>     • Flutter version 1.5.9-pre.46 at /Users/miottoan/flutter
>     • Framework revision e62b06a3ea (27 minutes ago), 2019-04-29 05:58:18 -0400
>     • Engine revision 48a5ce87ae
>     • Dart version 2.3.0 (build 2.3.0-dev.0.3 c46deebfb6)
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
>     • Android SDK at /Users/miottoan/Library/Android/sdk
>     • Android NDK location not configured (optional; useful for native profiling support)
>     • Platform android-28, build-tools 28.0.3
>     • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
>     • All Android licenses accepted.
> 
> [✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 10.2.1, Build version 10E1001
>     • ios-deploy 1.9.4
>     • CocoaPods version 1.6.0
> 
> [✓] Android Studio (version 3.4)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 35.0.1
>     • Dart plugin version 183.5901
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
> 
> [✓] VS Code (version 1.33.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 2.26.0
> 
> [✓] Connected device (1 available)
>     • iPhone Xʀ • 28F1BE2E-AE5F-4CAF-A1D8-0FBD47BA45FA • ios • com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

**Comment by HansMuller** [comment]
CC @gspencergoog

**Comment by gspencergoog** [comment]
Yes, this is due to a breaking change in `FocusScopeNode`.  Are you _hosting_ the focus node (calling `reparentIfNeeded`), or just creating one to pass in to a `TextField` or `FocusScope` widget?

**Comment by gspencergoog** [comment]
Actually, if you're calling `detach`, you're probably hosting the node.  There have been several changes in the way focus nodes work for widgets that are hosting nodes:

1) In order to host a node, you need to `attach` to the node in your `initState` method to create a [`FocusAttachment`](https://master-api.flutter.dev/flutter/widgets/FocusAttachment-class.html), passing in the context:
```dart
  FocusScopeNode _node;
  FocusAttachment _nodeAttachment;
  @override
  void initState() {
    super.initState();
    _node = FocusNode();
    _nodeAttachment = _node.attach(context);
  }
```
2) In your `dispose`, you need to call `dispose` on the focus scope node:
```dart
  @override
  void dispose() {
    // The attachment will automatically be detached in dispose().
    _node.dispose();
    super.dispose();
  }
```
3) In your build method, you will be calling `reparent` instead of `reparentIfNeeded`:
```dart
  @override
  Widget build(BuildContext context) {
    _nodeAttachment.reparent();
    return ...;
  }
```
4) Where you called `detach` before on the focus scope node, you will now call detach on the focus attachment.
```dart
  _nodeAttachment.detach();
```

However, I'd like to also point out that with the new focus API, it's less common to need to host a node yourself: you can just use the [`Focus`](https://master-api.flutter.dev/flutter/widgets/Focus-class.html) and [`FocusScope`](https://master-api.flutter.dev/flutter/widgets/FocusScope-class.html) widgets to handle all of the hosting mechanics (the things above), and simply pass the scope node to a `FocusScope` widget instead. You will still need to call `dispose` on a node that you own yourself. Even owning your own focus node isn't as necessary, unless you need an ancestor widget to be able to tell descendant nodes to be focused, since the `Focus` and `FocusScope` widgets will create and dispose of their own nodes if you don't supply one.

**Comment by AndreaMiotto** [comment]
It is still unclear to me how to close the keyboard on the textfield, at the moment I'm focusing a new FocusNode() like this:

`FocusScope.of(context).requestFocus(new FocusNode());`

But I know that this is not the proper way.

**Comment by gspencergoog** [comment]
You can use this: `FocusScope.of(context).unfocus()`
